cli/script/command-executor.ts

```diff
@@ -907,30 +908,66 @@ function getReactNativeProjectAppVersion(command: cli.IReleaseReactCommand, proj
             })
             .then((buildGradle: any) => {
-                if (buildGradle.android && buildGradle.android.defaultConfig && buildGradle.android.defaultConfig.versionName) {
-                    if (typeof buildGradle.android.defaultConfig.versionName !== "string") {
-                        throw new Error(`The "android.defaultConfig.versionName" property value in "android/app/build.gradle" is not a valid string. If this is expected, consider using the --targetBinaryVersion option to specify the value manually.`);
-                    }
+                if (!buildGradle.android || !buildGradle.android.defaultConfig || !buildGradle.android.defaultConfig.versionName) {
+                    throw new Error(`The "android/app/build.gradle" file doesn't specify a value for the "android.defaultConfig.versionName" property.`);
+                }
 
-                    var appVersion: string = buildGradle.android.defaultConfig.versionName.replace(/"/g, "").trim();
-                    if (semver.valid(appVersion) || missingPatchVersionRegex.test(appVersion)) {
-                        return appVersion;
-                    } else {
-                        throw new Error("The \"android.defaultConfig.versionName\" property in \"android/app/build.gradle\" needs to have at least a major and minor version, for example \"2.0\" or \"1.0.3\".");
-                    }
-                } else {
-                    throw new Error("The \"android/app/build.gradle\" file does not include a value for android.defaultConfig.versionName.");
+                if (typeof buildGradle.android.defaultConfig.versionName !== "string") {
+                    throw new Error(`The "android.defaultConfig.versionName" property value in "android/app/build.gradle" is not a valid string. If this is expected, consider using the --targetBinaryVersion option to specify the value manually.`);
+                }
+
+                let appVersion: string = buildGradle.android.defaultConfig.versionName.replace(/"/g, "").trim();
+
+                if (validVersion(appVersion)) {
+                    // The versionName property is a valid semver string,
+                    // so we can safely use that and move on.
+                    return appVersion;
+                } else if (Number(appVersion[0])) {
+                    // The versionName property isn't a valid semver string,
+                    // but it starts with a number, and therefore, it can't
+                    // be a valid Gradle property reference.
+                    throw new Error(`The "android.defaultConfig.versionName" property in "android/app/build.gradle" needs to specify a valid semver string, containing both a major and minor version (e.g. 1.3.2, 1.1).`);
                 }
+
+                // The version property isn't a valid semver string
+                // so we assume it is a reference to a property variable.
+                const propertyName = appVersion.replace("project.", "");
```

Comment:
My comment was overridden by the commit, re-posting:

Is it a given that the app version will always be referenced as project.APP_VERSION in a build gradle file? Can it not be something else eg: build.APP_VERSION or properties.APP_VERSION? How about just extracting the last property by splitting on .?